Add distributed mesh support and fake neighbor handling in BreakMeshByBlock

ChengHauYang opened the pull request:
closes #12033
closes #21161

## Reason

Currently, **BreakMeshByBlock (bmbb)** uses a hacky approach to prepare the mesh.  
When nodes are duplicated by bmbb, their former neighbors are no longer neighbors (since neighborhood is based on shared nodes). As a result:

- We cannot call `prepare_for_use()` at the end of bmbb, because it would erase the discontinuous neighbor information.
- To work around this, the framework has special hacks for bmbb (e.g., skipping mesh validity checks in debug mode when bmbb is present).

Another limitation is that bmbb cannot run with **distributed meshes**. This PR addresses that.

---

## Design

1. **Preserving discontinuous neighbors**  
   - Since `prepare_for_use()` wipes out discontinuous neighbor data, we introduce a new data structure to store this information (referred to as *fake neighbors*).  
   - These fake neighbors remain accessible to interface kernels that require them (e.g., CZM).

2. **Support for distributed meshes**  
   Following Roy’s suggestions:  
   - **Unique node IDs**: In replicated meshes, new nodes were created with  
     ```cpp
     Node::build(*current_node, mesh->n_nodes()).release();
     ```  
     This fails in distributed meshes.  
     -> Solution: use communication between processors and assign deterministic IDs based on `max_node_id` strides.  
   - **Ghost node handling**:  
     - *Phase 0*: Each rank computes locally connected blocks for its owned nodes and sends this info to the node owner.  
     - *Phase 1*: Node owners aggregate connected block info and broadcast the global set to all ranks.  
   - **Boundary ID assignment**: Updated in line with Roy’s fix ([commit link](https://github.com/roystgnr/moose/commit/2583950d46aa79a928903f392df908f4678f18d3)).

---

## Impact

- bmbb now supports **distributed meshes**.  
- `prepare_for_use()` can be safely called at the end of bmbb.  
- Removes the need for special hacks in the MOOSE framework for bmbb.

ChengHauYang committed:
Fix BreakMeshByBlockGenerator for distributed meshes and add FakeNeighbor RM

1. Distributed mesh fixes
   - Node ID assignment: using `mesh->n_nodes()` fails for distributed meshes (IDs are sparse). Switch to `max_node_id` with a stride.
   - Ghost node handling:
     - *Phase 0*: Each rank computes locally connected blocks for its owned nodes and sends them to the node owner.
     - *Phase 1*: Node owners aggregate connected block info and broadcast the global set to all ranks.
   - Boundary ID assignment aligned with Roy’s fix ([commit link](https://github.com/roystgnr/moose/commit/2583950d46aa79a928903f392df908f4678f18d3)).

2. New FakeNeighbor datastructure
   - Introduce a RelationshipManager for ghosted fake neighbors (CZM discontinuous neighbors, not topological).
   - Robust initialization:
     - Override `set_mesh` to ensure `_functor` is created JIT on first use.
     - Rebuild `_functor` in `internalInitWithMesh` if the neighbor map was populated after an early empty init.
     - Handles mesh cloning (`SetupMeshAction`) cases where `set_mesh` is called before initialization.

3. Simplifications
   - Since we now call `prepare_for_use()` at the end of BMBb, remove BreakMeshByBlockGenerator-specific hacks.
   - Simplify node ID handling.
   - Drop `MeshGeneratorSystem::_has_bmbb` flag and `hasBreakMeshByBlockGenerator()` API.

# Conflicts:
#	framework/src/base/MeshGeneratorSystem.C

# Conflicts:
#	framework/include/mesh/MooseMesh.h

ChengHauYang committed:
Always create map entries for nodes in _nodeid_to_connected_blocks

Fixes crashes from map::at when nodes had no valid blocks. Each node now
gets an entry (even empty), keeping the map consistent with node_to_elem_map.

ChengHauYang committed:
revert some changes so that the we do not create unwanted dulplicated nodes

ChengHauYang committed:
This commit fixes a segmentation fault when using block_pairs in distributed mode.

The crash was caused by an incorrect decision on node duplication, which stemmed from using incomplete, processor-local data (`node_to_elem_map`). The fix ensures the decision is based on globally synchronized data (`_nodeid_to_connected_blocks`), guaranteeing a consistent and correct outcome across all processors.

ChengHauYang committed:
If we set block_pairs in bmbb, the node should not duplicate when the node is in a junction with more than two blocks that share the node. It only happens that the node is in the interface of two blocks and has the block IDs equal to what the user set.

ChengHauYang committed:
closes #12033 and #21161 (1) remove REPLICATED mode restriction for bmbb tests (2) regold for the bmbb tests

hugary1995 commented on framework/include/meshgenerators/BreakMeshByBlockGeneratorBase.h:
```diff
@@ -37,12 +37,12 @@ class BreakMeshByBlockGeneratorBase : public MeshGenerator
-                              std::string & /*boundaryName*/,
-                              boundary_id_type & /*boundaryID*/,
-                              BoundaryInfo & /*boundary_info*/);
+  void findBoundaryName(MeshBase & mesh,
+                        const subdomain_id_type & /*primaryBlockID*/,
+                        const subdomain_id_type & /*secondaryBlockID*/,
+                        std::string & /*boundaryName*/,
+                        const boundary_id_type & /*boundaryID*/,
+                        BoundaryInfo & /*boundary_info*/);
```
```suggestion
  BoundaryID getBoundaryIDFromBlockPair(MeshBase & mesh,
                        const subdomain_id_type & /*primaryBlockID*/,
                        const subdomain_id_type & /*secondaryBlockID*/);
```

hugary1995 commented on framework/include/utils/MooseMeshUtils.h:
```diff
@@ -21,9 +21,9 @@ namespace MooseMeshUtils
 // Used to temporarily store information about which lower-dimensional
 // sides to add and what subdomain id to use for the added sides.
-struct ElemSidePair
+struct ElemIDSidePair
```
See if you are reinventing BndElement...

hugary1995 commented on framework/src/base/MeshGeneratorSystem.C:
```diff
@@ -43,9 +42,6 @@ MeshGeneratorSystem::addMeshGenerator(const std::string & type,
   if (_app.constructingMeshGenerators())
     createMeshGenerator(name);
-
-  if (type == "BreakMeshByBlockGenerator")
-    _has_bmbb = true;
```
💯

hugary1995 reviewed (changes requested):
Some early comments.

hugary1995 commented on framework/include/loops/ThreadedElementLoopBase.h:
```diff
@@ -276,7 +276,9 @@ ThreadedElementLoopBase<RangeType>::operator()(const RangeType & range, bool byp
             }
 
-          const Elem * neighbor = elem->neighbor_ptr(side);
+          const Elem * neighbor_ptr = elem->neighbor_ptr(side);
+          const Elem * neighbor =
+              (neighbor_ptr) ? neighbor_ptr : _mesh.neighbor_fake_ptr(elem, side);
```
Might have to add an assertion below to make sure it's nullptr.

ChengHauYang committed:
[WIP] change the neighbor ptr implementation to one line

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3267458/), step Clang format on 80ab17e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31636/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31636/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 23927d0dc7a66eccf59bbb36953ba4c8ff321a01`

ChengHauYang commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1434,6 +1408,44 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+    _elemid_side_to_fake_neighbor_elemid_side = std::move(map);
+  }
+
+  void clearElemSideToFakeNeighborElemSideMap()
+  {
+    _elemid_side_to_fake_neighbor_elemid_side.clear();
+  }
```
I just removed this member function.

ChengHauYang commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1434,6 +1408,44 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+    _elemid_side_to_fake_neighbor_elemid_side.clear();
+  }
+
+  const Elem * neighbor_fake_ptr(const Elem * elem, unsigned int side) const
```
I currently just rename it to be `fake_neighbor_ptr `. Let's wait for the suggestions.

ChengHauYang commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1434,6 +1408,44 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+      return nullptr;
+  }
+
+  int neighbor_fake_side(const Elem * elem, unsigned int side) const
```
Change it to be `fake_neighbor_side`

ChengHauYang commented on framework/include/meshgenerators/BreakMeshByBlockGenerator.h:
```diff
@@ -57,13 +58,40 @@ class BreakMeshByBlockGenerator : public BreakMeshByBlockGeneratorBase
+  typedef std::pair<dof_id_type, unsigned int> ElemIDSidePair;
 
-  std::set<std::pair<subdomain_id_type, subdomain_id_type>> _neighboring_block_list;
+  // a map from a node id to the set of connected block ids
+  std::map<dof_id_type, std::set<subdomain_id_type>> _nodeid_to_connected_blocks;
```
Thanks for pointing this out. I have updated the implementation so that `nodeid_to_connected_blocks` is now a local variable rather than a member variable.

ChengHauYang committed:
[WIP] API for fake neighbor

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3267719/), step Clang format on 92274e3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31636/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31636/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 23927d0dc7a66eccf59bbb36953ba4c8ff321a01`

hugary1995 committed:
slightly changed the API in MooseMesh related to disconnected neighbors

ChengHauYang committed:
Implement a parallel-safe disconnected neighbor system

Refactors the disconnected neighbor mechanism to be robust and safe under parallel (MPI) execution.

Previously, the implementation stored raw `Elem*` pointers in a map, which was fundamentally unsafe in a distributed environment. Pointers are only valid on the local rank and become meaningless across MPI processes, leading to segmentation faults in multi-process runs.

This commit resolves the issue by transitioning to an ID-based design:

- Core map (`_disconnected_neighbors_by_id`) now stores `dof_id_type` pairs instead of raw `ConstBndElement` pointers, ensuring portability and serializability across ranks.
- The `disconnectedNeighbor` accessor now returns a   `std::optional<ConstBndElement>`, replacing unsafe raw pointers with a safer C++ interface that naturally handles missing neighbors.

**Key Learning:** For parallel-safe design, always **store globally unique
IDs** and resolve them into rank-local pointers, which avoids pointer validity issues and ensures correctness across distributed
executions.

ChengHauYang committed:
clean the code

hugary1995 committed:
change data structure

ChengHauYang committed:
refactor(mesh): disconnected neighbor handling

- Replaced raw Elem* usage with ID-based ElemSide pairs for safer parallel handling
- Added cache (_cached_disconnected_neighbors) to speed up repeated lookups
- Modified neighbor queries across loops (NonlinearThread, ComputeUserObjectsThread, ComputeMaterialsObjectThread, FEProblemBase, etc.) to use disconnectedNeighborPtr

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3273454/), step Clang format on 0e481d2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31636/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31636/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b25e40313f46485fd1d3b2812d72e2c761e8427f`

ChengHauYang committed:
remove the place the do not need disconnectedNeighbor

ChengHauYang committed:
change libmesh side to reach disconnected neighbor through neighbor_ptr

ChengHauYang committed:
works for opt mode (not for dbg mode)

ChengHauYang committed:
update datastructure in libmesh

ChengHauYang committed:
support distributed meshes

ChengHauYang committed:
update libmesh to use disconnected boundary (libmesh commit: bee28f5)

ChengHauYang committed:
czm tests do not limit to replicated mesh only

ChengHauYang committed:
checkout libmesh

# Conflicts:
#	libmesh

ChengHauYang committed:
Add incomplete interface pair detection and CZM compatibility checks.

1. MooseMesh: Introduced `_has_incomplete_interface_pairs` flag with getter/setter, initialized in the constructor to detect missing reverse boundary pairs.
2. BreakMeshByBlockGenerator: Set `add_interface_on_two_sides = true` and `split_interface = true` by default; detects incomplete boundary pairs and outputs a warning if the user attempts to run CZM with such configurations.
3. CohesiveZoneAction & Tests: Added runtime error for incomplete CZM interfaces; updated test inputs to use explicit block-pair boundaries (replacing the unsupported single `interface` setup) and added the `boundary_pair_not_found` failure test

ChengHauYang committed:
update unit tests for new framework

ChengHauYang committed:
passing both replicated and distributed tests

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 910736a3a20aded528c3ba99371ce65980dc8e2c
+Subproject commit f956256a7b2b96dc0df2852e4f54f506ef20b062
```
**Caution!** This contains a submodule update

GiudGiud commented on framework/include/meshgenerators/BreakMeshByBlockGenerator.h:
```diff
@@ -59,11 +60,38 @@ class BreakMeshByBlockGenerator : public BreakMeshByBlockGeneratorBase
+  /// @brief a set of pairs of block ids between which new boundary sides are created
   std::set<std::pair<subdomain_id_type, subdomain_id_type>> _neighboring_block_list;
+
+  /// @brief a map from a pair of block ids to a set of element and side pairs
```
```suggestion
  /// @brief a map from a pair of block ids to a set of pairs of an element and a side
```

ChengHauYang replied:
Fixed.

GiudGiud commented on framework/include/meshgenerators/BreakMeshByBlockGenerator.h:
```diff
@@ -59,11 +60,38 @@ class BreakMeshByBlockGenerator : public BreakMeshByBlockGeneratorBase
+  std::unordered_map<std::pair<const Elem *, unsigned int>, std::pair<const Elem *, unsigned int>>
+      _elem_side_to_fake_neighbor_elem_side;
+
+  /// @brief Maps a pair of subdomain ids to the corresponding boundary id.
```
is that an extended map? that includes the disconnected / other-side of disconnection boundaries?

the MooseMesh has such a map already

ChengHauYang replied:
This was part of the legacy implementation. I have removed this member map in the latest commit.

GiudGiud commented on framework/include/relationshipmanagers/FakeNeighborRM.h:
```diff
@@ -0,0 +1,33 @@
+#include <unordered_map>
+#include "libmesh/elem.h"
+
+class FakeNeighborRM : public FunctorRelationshipManager
```
is fake better than disconnected here?

GiudGiud commented on modules/solid_mechanics/test/tests/cohesive_zone_model/poly.msh:
not used?

GiudGiud commented on modules/solid_mechanics/test/tests/cohesive_zone_model/czm_patch_test_base.i:
```diff
@@ -57,13 +57,13 @@
     variable = disp_y
     secondary = 'y1'    # boundary
-    penalty = 1e6
+    penalty = 5e6
```
these changes force you to regold

when chanigng the name of the interfaces likely do not because we dont check the sidesets in a regular exodiff (only check with -pedantic option)

ChengHauYang replied:
Thanks @GiudGiud! I reverted the penalty parameter change here.

I think it’s worth mentioning a few points:

1. I re-golded every file because the original `bmbb` did not call `prepare_for_use` at the end, which means it never went through `renumber_nodes_and_elements`. I have double-checked in ParaView that the solution matches the previous results within the specified tolerance.

2. However, we found that several CZM tests involving `EqualValueBoundaryConstraint` failed. The issue is that the original implementation selects the **node with the smallest ID** from a nodeset as the primary node. Since the mesh generator assigns different node numberings in replicated vs. distributed mode, the constraint ends up choosing different primary nodes, leading to small Exodus differences. To address this, we updated the `EqualValueBoundaryConstraint` to support specifying the primary node by coordinate, and added corresponding tests in `test/tests/constraints/equal_value_boundary_constraint/tests` to verify this new feature.

GiudGiud commented on framework/src/relationshipmanagers/FakeNeighborRM.C:
```diff
@@ -0,0 +1,129 @@
+  {
+    if (fn->map_empty() && !_elem_side_to_fake_neighbor_elem_side.empty())
+      _functor = std::make_unique<FakeNeighborFunctorImpl>(_elem_side_to_fake_neighbor_elem_side);
+  }
```
else assert the two maps are the same?

GiudGiud commented on framework/src/meshgenerators/BreakMeshByBlockGenerator.C:
```diff
@@ -136,210 +146,280 @@ BreakMeshByBlockGenerator::generate()
+              blockRestrictedElementSubdomainID(connected_elem);
 
-            if (current_elem != connected_elem && curr_elem_subid < connected_elem_subid)
+          if (current_elem != connected_elem && curr_elem_subid < connected_elem_subid)
```
add a comment

I d probably move current_elem != connected_elem higher to, with an explicit "continue;"

ChengHauYang replied:
Thank you, @GiudGiud! I added a higher early out.

GiudGiud reviewed (commented):
early review to keep up to date
looks good

GiudGiud commented on framework/src/meshgenerators/BreakMeshByBlockGenerator.C:
```diff
@@ -136,210 +146,280 @@ BreakMeshByBlockGenerator::generate()
+      fake_neighbor_rm->setFakeNeighborMap(elem_side_to_fake_neighbor_elem_side);
+    }
+    else
+      mooseError("Failed to cast to FakeNeighborRM.");
```
sounds good. feel free to copy paste that comment where needed and I ll resolve

ChengHauYang committed:
Framework: Coordinate-based EqualValueBoundaryConstraint node selection and removal of FakeNeighborRM

This update improves consistency and simplifies neighbor handling:

1. EqualValueBoundaryConstraint
   - Added `primary_node_coord` for selecting the primary node by coordinate, ensuring consistent behavior on distributed meshes where node IDs may differ.
   - If both `primary` and `primary_node_coord` are set, a `mooseError` is raised.
   - Added tests for coordinate-based selection and error handling.

2. Removed FakeNeighborRM
   - Eliminated the obsolete `FakeNeighborRM` used for ghosting in `BreakMeshByBlockGenerator`. libMesh ghosting now handles this.

ChengHauYang commented on framework/src/meshgenerators/BreakMeshByBlockGenerator.C:
```diff
@@ -136,210 +146,280 @@ BreakMeshByBlockGenerator::generate()
-          {
-            // assign the newly added node to current_elem
-            Node * new_node = nullptr;
+        if ((block_id != reference_subdomain_id) ||
```
Added. Thanks!

ChengHauYang commented on framework/src/relationshipmanagers/FakeNeighborRM.C:
```diff
@@ -0,0 +1,129 @@
+std::string
+FakeNeighborRM::getInfo() const
+{
+  return "FakeNeighborRM (ghosts across artificial neighbors)";
```
As mentioned above, I moved the ghost-element addition from MOOSE into libMesh.

ChengHauYang commented on framework/include/meshgenerators/BreakMeshByBlockGenerator.h:
```diff
@@ -59,11 +60,38 @@ class BreakMeshByBlockGenerator : public BreakMeshByBlockGeneratorBase
+           std::set<std::pair<const Elem *, unsigned int>>>
       _new_boundary_sides_map;
+
+  /// @brief Maps an element pointer and side pair to its corresponding fake neighbor element pointer and side pair.
```
Fixed "fake" to be "disconnected"

ChengHauYang commented on framework/include/relationshipmanagers/FakeNeighborRM.h:
```diff
@@ -0,0 +1,33 @@
+#pragma once
```
Fixed

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 910736a3a20aded528c3ba99371ce65980dc8e2c
+Subproject commit d615ffee1aca885edd568b59d1b8dd969eb434c2
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

hugary1995 commented:
@loganharbour I helped @ChengHauYang perform a libmesh package update in this PR. We want to see if our proposed libmesh change https://github.com/libMesh/libmesh/pull/4283 is sufficient for these moose changes. Do you need to trust some commits to activate the CI workflows?

ChengHauYang committed:
Strengthen EVBC node selection & validation

- Improved primary node detection:
  - Ensures node matches either in secondary IDs or belongs to secondary boundary.
  - Added errors for zero or multiple matches.
- Added test `error_multiple_primary_nodes` for duplicate primary nodes found error.

ChengHauYang committed:
fix the unit tests

ChengHauYang committed:
checkout libmesh to devel

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 910736a3a20aded528c3ba99371ce65980dc8e2c
+Subproject commit 6cd8e2172ba26c42696d61947bac69ebeae5c58c
```
**Caution!** This contains a submodule update

ChengHauYang committed:
slighly clean the unused functions

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3449666/), step Clang format on 32a6711 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31636/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31636/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c5e21738cfb50eb1f079309dc5d52ed59c16965d`

ChengHauYang committed:
add back missing APIs after rebasing

ChengHauYang committed:
regold one file

ChengHauYang committed:
Ensure unique global primary-node selection under distributed meshes

Primary-node detection previously relied on summing local matches across ranks,
which incorrectly counted shared boundary nodes multiple times. This patch
replaces rank-local counting with a global deduplication of matched node IDs:
each rank reports its unique local matches, all IDs are gathered, a global
unique set is formed, and the smallest ID is chosen to ensure deterministic
selection. This guarantees correct behavior regardless of mesh partitioning.
